In [1]:
import pandas as pd
import keras
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
import numpy as np

Using TensorFlow backend.


In [2]:
df = pd.read_csv('TensorData.csv')

df.describe()

Covert Index  Number of CCs   Corelation  Conspirator
count   3961.000000    3962.000000  3962.000000  3962.000000
mean       1.544912      26.661282     0.227096     0.045179
std        0.579729      75.551440     0.394869     0.207723
min        0.000000       0.000000   -20.243486     0.000000
25%        1.115798       0.000000     0.159040     0.000000
50%        1.506680       4.000000     0.301030     0.000000
75%        1.977152      24.000000     0.301030     0.000000
max        3.068469    1162.000000     2.107210     1.000000

In [3]:
df = df[df['Number of CCs'] < 1000]

In [4]:
features = ['Covert Index','Number of CCs','Corelation']

target = ['Conspirator']

# Preprocessing

Normalize between -1 and 1

fill na

In [5]:
df = df.fillna(0)

In [6]:
df.head()

Covert Index  Number of CCs  Corelation                   Email 1  \
0      2.817695             48   -1.505150      tana.jones@enron.com   
1      1.278754              0   -0.477121  richard.causey@enron.com   
2      2.890906             80    0.139794     steven.kean@enron.com   
3      2.890906             80    0.139794     steven.kean@enron.com   
4      2.115726             42    0.150515   mark.haedicke@enron.com   

                     Email 2  Conspirator  
0   louise.kitchen@enron.com            1  
1       sally.beck@enron.com            1  
2         ken.rice@enron.com            1  
3            skean@enron.com            1  
4  richard.sanders@enron.com            1

In [7]:
for feature in features:
    df[feature] = (df[feature] - df[feature].mean())/df[feature].std()

In [8]:
df[features]

Covert Index  Number of CCs  Corelation
0         2.203561       0.488971   -7.787760
1        -0.454899      -0.480427   -3.181219
2         2.330029       1.135236   -0.416856
3         2.330029       1.135236   -0.416856
4         0.990937       0.367796   -0.368815
5        -0.516115      -0.440035   -0.368815
6        -0.516115      -0.440035   -0.368815
7         2.203561       0.488971   -0.330613
8         0.882716      -0.480427   -0.330613
9        -0.495462      -0.480427    0.119020
10        0.300432      -0.480427    0.305634
11        1.248487      -0.480427    0.305634
12        2.289209       1.094845    0.305634
13        0.883821       0.004272    0.305634
14       -0.338241      -0.359252    0.305634
15        1.305115       3.437557    0.305634
16        0.300432      -0.480427    0.305634
17        0.883821       0.004272    0.305634
18        2.203561       0.488971    0.305634
19       -1.204022      -0.480427    0.305634
20       -1.204022      -0.480427    0.305634
21       -1.204022      -0.480427    0.305634
22        1.937267       1.014062    0.305634
23        2.203561       0.488971    0.305634
24        2.203561       0.488971    0.305634
25        2.203561       0.488971    0.305634
26        0.990937       0.367796    0.980084
27       -0.279635      -0.480427    1.094689
28       -1.204022      -0.480427    1.094689
29        0.907132      -0.399644    1.094689
...            ...            ...         ...
3932     -0.799652      -0.480427    0.305634
3933     -1.623860      -0.480427    0.305634
3934      1.171464       0.408188    0.305634
3935      0.414134      -0.399644    0.305634
3936      0.414134      -0.399644    0.305634
3937      0.414134      -0.399644    0.305634
3938      0.912406       3.720298    0.305634
3939      0.912406       3.720298    0.305634
3940      0.003421      -0.480427    0.305634
3941      0.174220       0.933279    0.305634
3942      1.603011       4.528129    0.305634
3943      0.696154      -0.318860    0.305634
3944      0.669099      -0.480427    0.305634
3945      1.211308      -0.318860    0.305634
3946     -0.583826      -0.480427    0.305634
3947     -1.421067      -0.399644    0.305634
3948     -1.421067      -0.399644    0.305634
3949      0.964411      -0.480427    0.305634
3950     -1.623860      -0.480427    0.305634
3951     -1.623860      -0.480427    0.305634
3952     -1.623860      -0.480427    0.305634
3953     -0.495462      -0.480427    0.305634
3954     -0.495462      -0.480427    0.305634
3955     -0.495462      -0.480427    0.305634
3956      0.696154      -0.318860    0.305634
3957     -0.799652      -0.480427    0.305634
3958      0.485317       0.085055    0.305634
3959     -0.198577      -0.318860    0.305634
3960      0.901658       1.135236    0.305634
3961     -0.301415      -0.399644    0.305634

[3952 rows x 3 columns]

In [9]:
X = df[:1000][features].values.astype(np.float64)
y = df[:1000][target].values.astype(np.float64)

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=77)

In [11]:
model = keras.Sequential()
model.add(keras.layers.Dense(12,input_dim=(3) ,activation='relu'))
model.add(keras.layers.Dense(5, activation='relu'))
model.add(keras.layers.Dense(1, activation='sigmoid'))

In [12]:
model.compile(loss='binary_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

Accuracy to beat:

In [13]:
100-17.9

82.1

# Model 1:  Without any regularization

Accuracy with random choosing = 82.1 %

In [14]:
model.fit(X_train, y_train, epochs=400, validation_data=(X_test, y_test), batch_size=50)

Train on 670 samples, validate on 330 samples
Epoch 1/400
670/670 [==============================] - 0s 505us/step - loss: 0.9074 - acc: 0.1701 - val_loss: 0.8740 - val_acc: 0.1333
Epoch 2/400
670/670 [==============================] - 0s 44us/step - loss: 0.8418 - acc: 0.1896 - val_loss: 0.8288 - val_acc: 0.2121
Epoch 3/400
670/670 [==============================] - 0s 41us/step - loss: 0.7947 - acc: 0.3642 - val_loss: 0.7893 - val_acc: 0.3273
Epoch 4/400
670/670 [==============================] - 0s 44us/step - loss: 0.7545 - acc: 0.4164 - val_loss: 0.7562 - val_acc: 0.3848
Epoch 5/400
670/670 [==============================] - 0s 47us/step - loss: 0.7228 - acc: 0.4731 - val_loss: 0.7294 - val_acc: 0.4545
Epoch 6/400
670/670 [==============================] - 0s 43us/step - loss: 0.6974 - acc: 0.5522 - 

In [15]:
y_pred = model.predict(X_test)

# [true negative, false positive, false negative, true positive]

In [16]:
y_pred.round(decimals=2)

array([[1.  ],
       [0.97],
       [0.06],
       [0.02],
       [0.01],
       [1.  ],
       [0.05],
       [0.02],
       [0.01],
       [0.  ],
       [0.79],
       [0.02],
       [0.  ],
       [0.04],
       [0.02],
       [0.94],
       [0.09],
       [1.  ],
       [0.02],
       [0.06],
       [0.04],
       [0.08],
       [0.  ],
       [0.02],
       [1.  ],
       [0.03],
       [0.03],
       [0.02],
       [0.02],
       [0.02],
       [0.02],
       [0.04],
       [1.  ],
       [0.11],
       [0.04],
       [0.02],
       [0.  ],
       [0.  ],
       [1.  ],
       [0.05],
       [0.94],
       [0.07],
       [0.01],
       [0.02],
       [0.04],
       [0.02],
       [0.15],
       [0.09],
       [1.  ],
       [0.08],
       [0.01],
       [0.07],
       [0.02],
       [1.  ],
       [0.02],
       [1.  ],
       [0.05],
       [0.01],
       [1.  ],
       [1.  ],
       [0.02],
       [1.  ],
       [0.02],
       [0.02],
       [0.02],
       [0.01],
       [0.

In [17]:
confusion_matrix(y_test, (y_pred > 0.5).astype('float')).ravel()

array([264,   1,  11,  54])

total test dataset size is

In [18]:
confusion_matrix(y_train, (model.predict(X_train)>0.5)).ravel()

array([554,   2,  25,  89])

## lots of false negatives. Very bad

# Accuracy

In [19]:
np.mean(y_test == (y_pred>0.5)) * 100

96.36363636363636

# With class regularization

In [20]:
model.fit(X_train, y_train, epochs=400, validation_data=(X_test, y_test), batch_size=50, class_weight={0:0.179, 1:1})

Train on 670 samples, validate on 330 samples
Epoch 1/400
670/670 [==============================] - 0s 43us/step - loss: 0.0931 - acc: 0.9627 - val_loss: 0.1294 - val_acc: 0.9667
Epoch 2/400
670/670 [==============================] - 0s 40us/step - loss: 0.0841 - acc: 0.9552 - val_loss: 0.1324 - val_acc: 0.9606
Epoch 3/400
670/670 [==============================] - 0s 40us/step - loss: 0.0807 - acc: 0.9478 - val_loss: 0.1376 - val_acc: 0.9576
Epoch 4/400
670/670 [==============================] - 0s 47us/step - loss: 0.0789 - acc: 0.9388 - val_loss: 0.1439 - val_acc: 0.9515
Epoch 5/400
670/670 [==============================] - 0s 42us/step - loss: 0.0780 - acc: 0.9358 - val_loss: 0.1474 - val_acc: 0.9515
Epoch 6/400
670/670 [==============================] - 0s 43us/step - loss: 0.0776 - acc: 0.9373 - 

In [21]:
y_pred = model.predict(X_test)

In [22]:
from sklearn.metrics import confusion_matrix

# [true negative, false positive, false negative, true positive]

In [23]:
confusion_matrix(y_test, (y_pred > 0.5).astype('float')).ravel()

array([258,   7,  10,  55])

total test dataset size is

In [24]:
len(y_pred)

330

In [25]:
confusion_matrix(y_train, (model.predict(X_train)>0.5)).ravel()

array([536,  20,  11, 103])

# Accuracy

In [26]:
np.mean(y_test == (y_pred>0.5)) * 100

94.84848484848484

## Much fewer False Negatives. 